docs/source/python/integration/dataset.rst

```diff
@@ -0,0 +1,134 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+Extending PyArrow Datasets
+==========================
+
+PyArrow provides a core protocol for datasets, so third-party libraries can both
+produce and consume classes that conform to useful subset of the PyArrow dataset
+API. This subset provides enough functionality to provide predicate and filter
+pushdown. The subset of the API is contained in ``pyarrow.dataset.protocol``.
+
+.. image:: pyarrow_dataset_protocol.svg
+   :alt: A diagram showing the workflow for using the PyArrow Dataset protocol.
+         There are two flows shown, one for stream and one for tasks. The stream
+         case shows a linear flow from a producer class, to a dataset, to a 
+         scanner, and finally to a RecordBatchReader. The tasks case shows a
+         similar diagram, except the dataset is split into fragments, which are
+         then distributed to tasks, which each create their own scanner and
+         RecordBatchReader.
+
+Producers are responsible for outputting a class that conforms to the protocol.
+
+Consumers are responsible for calling methods on the protocol to get the data
+out of the dataset. The protocol supports getting data as a single stream or
+as a series of tasks which may be distributed.
+
+From the perspective of a user, the code looks like:
+
+.. code-block:: python
+
+    dataset = producer_library.get_dataset(...)
+    df = consumer_library.read_dataset(dataset)
+    df.filter("x > 0").select("y")
```

Comment:
I'm confused about `select()` here. The Dataset API does not currently expose a `select()` method, but it allows column projection by allowing users to pass column names or expressions to other methods or to the scanner. Do you propose adding a `select()` method to the Dataset API?

Reply:
This is meant to be a sketch of what usage looks like from the users perspective. `df` is a DataFrame object produced by an imaginary consumer library, and `filter` and `select` are methods on this imaginary library.

The idea is the user gets a dataset from the producer, and passes it to the consumer. Then the user gets to write their queries with their favorite query engine, and under the hood the datasets API is used to do projection and predicate pushdown.

Reply:
Perhaps I should explain that it's imaginary? Or I could show a concrete usage with DuckDB and then Polars?

Reply:
I've switched this to a concrete example with Delta Lake and DuckDB.